lib/Service/ImportService.php

```diff
@@ -57,6 +61,124 @@ public function __construct(PermissionsService $permissionsService, LoggerInterf
 		$this->userManager = $userManager;
 	}
 
+	public function previewImport(?int $tableId, ?int $viewId, string $path): array {
+		if ($viewId !== null) {
+			$this->viewId = $viewId;
+		} elseif ($tableId) {
+			$this->tableId = $tableId;
+		} else {
+			$e = new \Exception('Neither tableId nor viewId is given.');
+			$this->logger->error($e->getMessage(), ['exception' => $e]);
+			throw new InternalError(get_class($this) . ' - ' . __FUNCTION__ . ': '.$e->getMessage());
+		}
+
+		$this->createUnknownColumns = false;
+		$previewData = [];
+
+		try {
+			$userFolder = $this->rootFolder->getUserFolder($this->userId);
+			$error = false;
+			if ($userFolder->nodeExists($path)) {
+				$file = $userFolder->get($path);
+				$tmpFileName = $file->getStorage()->getLocalFile($file->getInternalPath());
+				if($tmpFileName) {
+					$spreadsheet = IOFactory::load($tmpFileName);
+					$previewData = $this->getPreviewData($spreadsheet->getActiveSheet());
+				} else {
+					$error = true;
+				}
+			} elseif (\file_exists($path)) {
+				$spreadsheet = IOFactory::load($path);
+				$previewData = $this->getPreviewData($spreadsheet->getActiveSheet());
+			} else {
+				$error = true;
+			}
+
+			if($error) {
+				throw new NotFoundError('File for import could not be found.');
+			}
+
+		} catch (NotFoundException|NotPermittedException|NoUserException|InternalError|PermissionError $e) {
+			$this->logger->warning('Storage for user could not be found', ['exception' => $e]);
+			throw new NotFoundError('Storage for user could not be found');
+		}
+
+		return $previewData;
+	}
+
+	/**
+	 * @param Worksheet $worksheet
+	 * @throws DoesNotExistException
+	 * @throws InternalError
+	 * @throws MultipleObjectsReturnedException
+	 * @throws NotFoundError
+	 * @throws PermissionError
+	 */
+	private function getPreviewData(Worksheet $worksheet): array {
+		$firstRow = $worksheet->getRowIterator()->current();
+		$secondRow = $worksheet->getRowIterator()->seek(2)->current();
+
+		// Prepare columns data
+		$columns = [];
+		$this->getColumns($firstRow, $secondRow);
+
+		foreach ($this->rawColumnTitles as $colIndex => $title) {
+			if ($this->columns[$colIndex] !== '') {
+				/** @var Column $column */
+				$column = $this->columns[$colIndex];
+				$columns[] = $column;
+			} else {
+				$columns[] = [
+					'title' => $title,
+					'type' => $this->rawColumnDataTypes[$colIndex]['type'],
+					'subtype' => $this->rawColumnDataTypes[$colIndex]['subtype'],
+					'numberDecimals' => $this->rawColumnDataTypes[$colIndex]['number_decimals'] ?? 0,
+					'numberPrefix' => $this->rawColumnDataTypes[$colIndex]['number_prefix'] ?? '',
+				];
+			}
+		}
+
+		// Prepare rows data
+		$count = 0;
+		$maxCount = 3;
+		$rows = [];
+
+		foreach ($worksheet->getRowIterator(2) as $row) {
+			$rowData = [];
+			$cellIterator = $row->getCellIterator();
+			$cellIterator->setIterateOnlyExistingCells(false);
+
+			foreach ($cellIterator as $cellIndex => $cell) {
```

Comment:
static analysis looks like you need to specify that `$cellIndex` is an integer here.
Maybe you can specify `$cellIterator` to be an [iterable](https://psalm.dev/docs/annotating_code/type_syntax/other_types/) over ints or you specify `$cellIndex` to be an `int<0, max>`.


```